crates/sui-rpc-api/proto/sui/rpc/v2alpha/move_descriptor.proto

```diff
@@ -0,0 +1,210 @@
+  optional uint32 position = 2;
+
+  // The type of the field
+  optional OpenSignatureBody type = 5;
+}
+
+enum Type {
+  TYPE_UNKNOWN = 0;
+  ADDRESS = 1;
+  BOOL = 2;
+  U8 = 3;
+  U16 = 4;
+  U32 = 5;
+  U64 = 6;
+  U128 = 7;
+  U256 = 8;
+  VECTOR = 9;
+  DATA_TYPE = 10;
+  TYPE_PARAMETER = 11;
+}
+
+// Descriptor of an enum variant
+message VariantDescriptor {
+  // Name of the variant
+  optional string name = 1;
+
+  // Order or position of the variant in the enum definition.
+  optional uint32 position = 2;
+
+  // Set of fields defined by this variant.
+  repeated FieldDescriptor fields = 3;
+}
+
+message OpenSignatureBody {
+  // Type of this signature
+  optional Type type = 1;
+
+  // position of the type parameter as defined in the containing data type descriptor when `type` is `TYPE_PARAMETER`
+  optional uint32 type_parameter = 2;
+
+  // Fully qualified name of the data type when `type` is `DATA_TYPE`
+  optional string type_name = 3;
+
+  // Set when `type` is `VECTOR` or `DATA_TYPE`
+  repeated OpenSignatureBody type_parameter_instantiation = 4;
+}
+
+// Descriptor of a Move function
+message FunctionDescriptor {
+  // Name of the function
+  optional string name = 1;
+
+  // XXX Do we want both storage/runtime info here??
+  // PackageId where this function is defined.
+  optional string package_id = 2;
+
+  // Name of the module where this function is defined
+  optional string module = 3;
+
+  // Fully qualified name of this function.
+  //
+  // This is `<package_id>::<module>::<name>`
+  optional string function_name = 4;
+  
+  // Whether the function is `public`, `private` or `public(friend)`
+  optional Visibility visibility = 5;
+
+  // Whether the function is marked `entry` or not.
+  optional bool is_entry = 6;
+
+  // Ability constraints for type parameters
+  repeated Ability type_parameters = 7;
+
+  // Formal parameter types.
+  repeated OpenSignature parameters = 8;
+
+  // Return types.
+  repeated OpenSignature returns = 9;
+}
+
+enum Reference {
+  REFERENCE_UNKNOWN = 0;
+  IMMUTABLE = 1;
+  MUTABLE = 2;
+}
+
+enum Visibility {
+  VISIBILITY_UNKNOWN = 0;
+  PRIVATE = 1;
+  PUBLIC = 2;
+  FRIEND = 3;
+}
+
+message OpenSignature {
+  optional Reference reference = 1;
+  optional OpenSignatureBody body = 2;
+}
+
+message Module {
+  // Name of this module
+  optional name = 1;
+
+  // PackageId of the package where this module is defined.
+  optional package_id = 2;
+
+  // List of DataTypes defined by this module.
+  repeated Descriptor data_types = 3;
+
+  // List of Functions defined by this module.
+  repeated FunctionDescriptor functions = 4;
+
+  // List of Constants defined by this module.
+  // repeated Constant constants = 5;
+}
+
+// A Move Package
+message Package {
+  // The PackageId of this package
+  // The Id this package was loaded from on-chain.
+  optional string package_id = 1;
+
+  // XXX Do we need/what to distinguish this?
```

Comment:
I would probably prefer this. But I'm also kinda weird in that I prefer these things to be very explicit.